packages/shared/src/features/annotation/types.ts

```diff
@@ -1,15 +1,97 @@
 import z from "zod";
-import { ScoreConfig } from "../../db";
+import { type ScoreDataType, type ScoreSource } from "../../db";
 
-const configCategory = z.object({
+export const Category = z.object({
   label: z.string().min(1),
   value: z.number(),
 });
 
-export const categoriesList = z.array(configCategory);
+export type ConfigCategory = z.infer<typeof Category>;
 
-export type ConfigCategory = z.infer<typeof configCategory>;
+const NUMERIC: ScoreDataType = "NUMERIC";
+const CATEGORICAL: ScoreDataType = "CATEGORICAL";
+const BOOLEAN: ScoreDataType = "BOOLEAN";
 
-export type CastedConfig = Omit<ScoreConfig, "categories"> & {
-  categories: ConfigCategory[] | null;
-};
+const API: ScoreSource = "API";
+const EVAL: ScoreSource = "EVAL";
+const ANNOTATION: ScoreSource = "ANNOTATION";
+
+export const availableDataTypes = [NUMERIC, CATEGORICAL, BOOLEAN] as const;
+const availableSources = [API, EVAL, ANNOTATION] as const;
+
+export const createConfigSchema = z.object({
+  name: z.string().min(1).max(35),
+  dataType: z.enum(availableDataTypes),
+  minValue: z.coerce.number().optional(),
+  maxValue: z.coerce.number().optional(),
+  categories: z.array(Category).optional(),
+  description: z.string().optional(),
+});
+
+export type CreateConfig = z.infer<typeof createConfigSchema>;
+
+const NumericData = z.object({
+  value: z.number(),
+  stringValue: z.undefined().nullish(),
+  dataType: z.literal("NUMERIC"),
+});
+
+const CategoricalData = z.object({
+  value: z.number().optional().nullish(),
+  stringValue: z.string(),
+  dataType: z.literal("CATEGORICAL"),
+});
+
+const BooleanData = z.object({
+  value: z.number(),
+  stringValue: z.string(),
+  dataType: z.literal("BOOLEAN"),
+});
+
+const ScoreBase = z.object({
+  id: z.string(),
+  timestamp: z.coerce.date(),
+  projectId: z.string(),
+  name: z.string(),
+  source: z.enum(availableSources),
+  authorUserId: z.string().nullish(),
+  comment: z.string().nullish(),
+  traceId: z.string(),
+  observationId: z.string().nullish(),
+  configId: z.string().nullish(),
+  createdAt: z.coerce.date(),
+  updatedAt: z.coerce.date(),
+});
+
+export const ValidatedScoreSchema = z.discriminatedUnion("dataType", [
+  ScoreBase.merge(NumericData),
+  ScoreBase.merge(CategoricalData),
+  ScoreBase.merge(BooleanData),
+]);
+
+export type ValidatedScore = z.infer<typeof ValidatedScoreSchema>;
```

Comment:
Note: this currently is defined once for public api and once here for langfuse codebase use. Is there a better way to do this? Ideally I'd like to share this type across both use cases, to ensure we always expect scores objects to have this data type